File: dump-cn/20.valid-parentheses.cs
public class Solution {
    public bool IsValid(string s)
    {
        int m = s.Length;
        if(m == 0)
            return true;
        if (m < 2 || m % 2 == 1)
            return false;
        int i = 0;
        Stack<char> s1 = new Stack<char>();
        while (i < m)
        {
            switch (s[i])
            {
                case '(':
                    s1.Push(s[i]);
                    break;
                case '[':
                    s1.Push(s[i]);
                    break;
                case '{':
                    s1.Push(s[i]);
                    break;
                case ')':
                    if (s1.Count != 0 && s1.Peek() == '(')
                        s1.Pop();
                    else
                        return false;
                    break;
                case ']':
                    if (s1.Count != 0 && s1.Peek() == '[')
                        s1.Pop();
                    else
                        return false;
                    break;
                case '}':
                    if (s1.Count != 0 && s1.Peek() == '{')
                        s1.Pop();
                    else
                        return false;
                    break;
            }
            ++i;
        }
        return s1.Count == 0;
    }
}

File: dump-cn/20.valid-parentheses.cpp
class Solution {
public:
 	bool isValid(string s) {
		stack<char> st;
		int index = 0;
		while (index < s.length()) {
			switch (s[index])
			{
			case '(':
				st.push(s[index]);
				break;
			case '[':
				st.push(s[index]);
				break;
			case '{':
				st.push(s[index]);
				break;
			case ')':
				if (st.empty() || st.top() != '(')
					return false;
				else
					st.pop();
				break;
			case ']':
				if (st.empty() || st.top() != '[')
					return false;
				else
					st.pop();				
				break;
			case '}':
				if (st.empty() || st.top() != '{')
					return false;
				else
					st.pop();				
				break;
			default:
				return false;
			}
			index++;
		}

		return st.empty();
	}
};